Mu2eG4/src/SimParticlePrimaryHelper.cc

```diff
@@ -1,30 +1,72 @@
 // Andrei Gaponenko, 2013
 
 #include "Mu2eG4/inc/SimParticlePrimaryHelper.hh"
-#include "art/Framework/Principal/Event.h"
+
+#include "MCDataProducts/inc/StepPointMC.hh"
+#include "MCDataProducts/inc/StageParticle.hh"
+
+#include "cetlib_except/exception.h"
+
+#include <iostream>
+#define AGDEBUG(stuff) std::cerr<<"AG: "<<__FILE__<<", line "<<__LINE__<<": "<<stuff<<std::endl;
 
 namespace mu2e {
-  SimParticlePrimaryHelper::SimParticlePrimaryHelper(const art::Event* event,
-                                                     const art::ProductID& simProdID,
+
+  //================================================================
+  SimParticlePrimaryHelper::SimParticlePrimaryHelper(const art::ProductID& simProdID,
                                                      const art::EDProductGetter* sim_prod_getter):
     simProdID_(simProdID),
-    event_(event),
     simProductGetter_(sim_prod_getter)
   {}
 
-  void SimParticlePrimaryHelper::addEntryFromGenParticle(const art::ValidHandle<GenParticleCollection>& gensHandle, unsigned genId)
-  {
-    entries_.emplace_back(art::Ptr<GenParticle>(gensHandle, genId),
-                          art::Ptr<SimParticle>(simProdID_) );
+  //================================================================
+  art::Ptr<GenParticle> SimParticlePrimaryHelper::genParticlePtr(int g4TrkID) const {
+    art::Ptr<GenParticle> res;
+    auto pgen = std::get_if<art::Ptr<GenParticle> >(&entries_.at(g4TrkID - 1));
```

Comment:
why don't you use getEntry here?  That would make the code more consistent and easier to understand.

Reply:
Because that code was written  before getEntry() was introduced.   Also getEntry() returns an rvalue, while we need an lvalue for the get_if call.  However invoking getEntry does make the code cleaner, so I made the switch as the price of an extra copy.  I think it is worth it.   94a6ee529ac46d1360e8424c0db8e9414341225f